### ISSUE TITLE
Should LiveView provide equivalent components to Phoenix.HTML functions?

### ISSUE THREAD

**Issue by thiagomajesk** [phoenixframework/phoenix_live_view#2409]
Hi!

I've been testing Phoenix's 1.7 release candidates for a while and I see that there's quite a bit of code generated for the new core components module, especially regarding forms. We now have a good chunk of logic to maintain what was previously abstracted by `Phoenix.HTML` (still is...). For instance, core components now call some functions directly: `Phoenix.HTML.Form.input_name/2`, `Phoenix.HTML.Form.input_id/2`, `Phoenix.HTML.Form.input_value/2`, `Phoenix.HTML.Form.options_for_select/2`, `input_equals`, etc.

I'm not quite sure what are the plans for `Phoenix.HTML` and `Phoenix.Component`, but I see that the intersection between both APIs is growing. Just recently we got `inputs_for` added to `Phoenix.Component` so I was thinking if we could get an `input` component as well, just to remove some of the maintenance burden.

I know that the idea with the new generators is kinda "flexibility over abstraction", but for this particular case, I think that it might be better to have form-related logic inside the framework. I'm worried that since this implementation is being pushed as part of the development process, I'll have to keep track of changes and improvements to the framework implementation. For example, this recently merged [commit](https://github.com/phoenixframework/phoenix/commit/b3db2e7b876ed58ddbcb62a4e90cf3986d869874) fixes a problem with how dates were handled, which means I'd have to manually 'upgrade' the implementation in all projects.

Perhaps a solution could be adding a new `input` component to `Phoenix.Component` that has the core logic and then simplifying the generated implementation!? I'm not sure... Ideally, I'd like to be able to use all the functions from `Phoenix.HTML` (with and without forms) as components (`checkbox`, `button`, `email_input`, `file_input`, `button`, etc).

Does this make sense to someone else or is it just a minor annoyance I'm having with the prospect of maintaining all this generated code in the future?

**Comment by josevalim** [comment]
One important thing to keep in mind is that the inputs in your generated app are not the same than the ones in your Phoenix.HTML. For example, we can't abstract all of the markup. Take the regular input component:

```elixir
    <div phx-feedback-for={@name}>
      <.label for={@id}><%= @label %></.label>
      <input
        type={@type}
        name={@name}
        id={@id || @name}
        value={@value}
        class={[
          input_border(@errors),
          "mt-2 block w-full rounded-lg border-zinc-300 py-[7px] px-[11px]",
          "text-zinc-900 focus:outline-none focus:ring-4 sm:text-sm sm:leading-6",
          "phx-no-feedback:border-zinc-300 phx-no-feedback:focus:border-zinc-400 phx-no-feedback:focus:ring-zinc-800/5"
        ]}
        {@rest}
      />
      <.error :for={msg <- @errors}><%= msg %></.error>
    </div>
```

If you were to abstract this into an `.input` helper that is part of Phoenix, the only difference is that you would write:

```
<.input
  type={@type}
  form={f}
  field={field}
```

instead of:

```
<input
  type={@type}
  name={@name}
  id={@id || @name}
  value={@value}
```

This is in itself is not a huge difference and, staying closer to the markup gives much more control for you to customize it in your own apps.

However, I agree with you that there is some shared logic/boilerplate. I can identify three:

1. format_input_value (which you added in your PR)
2. input_equals?
3. the calls to input_value/input_name/input_id

So I'd suggest to focus on eliminating those instead.

For example, maybe we could introduce a `Phoenix.HTML.Form.format_value(type, value)`, which returns the value properly formatted according to the input. This would remove the need for your PR (which will cast naive/datetime for "datetime-local") and also fix `input_equals?`, because we could make it so `format_value("checkbox", value)` returns either `true | false`.

I am not sure how to tackle the third one, but I believe we will only have the answer when we finish work on the new forms. Would you like to send a PR for Phoenix.HTML that adds this new function to core components?

**Comment by thiagomajesk** [comment]
> One important thing to keep in mind is that the inputs in your generated app are not the same than the ones in your Phoenix.HTML. For example, we can't abstract all of the markup

Yeah, yeah, of course... From reading your reply I think you already understood most of what my concern is coming from. Let's see if we can figure out how to improve on those points. 

1. As you said, we could add a `format_value` helper to `Phoenix.HTML.Form`. Another option that would work well is to move the formatting logic directly into the `input_value` function. I think it's reasonable to assume that when you call `input_value`, you expect to have a valid value to use on the input and not just extract it from the form source (I'm also unsure what would be the use cases for a `format_value` function outside of this particular use-case).

2. "input_equals?" -> Probably, but should this be a public function? It'll depend on how we deal with items nº 1 and 3 (and how much we want to expose to the end-user).

3. If we break the current generated `.input` helper apart, we can see that it's actually a group of label + input + error (perhaps we should even rename it to `.input_group`)¹. Based on that, I think we could extract three components to `Phoenix.Component`: `.input`, `.label`, and `.error`, leaving our core components with a simpler implementation, a single component:

```elixir
def input_group(assigns) do
  ~H"""
    <div phx-feedback-for={@name}>
      <.label form={f} field={field} class="block text-sm font-semibold leading-6 text-zinc-800" />
      <.input type={@type} form={f} field={field} class={input_classes(@type)} {@rest} />
      <.errors let={error} form={f} field={field} translator={&translate_errors} class="phx-no-feedback:hidden mt-3 flex gap-3 text-sm leading-6 text-rose-600">
        <Heroicons.exclamation_circle mini class="mt-0.5 h-5 w-5 flex-none fill-rose-500" />
        <%= error %>
      </.errors>
    </div>
    """
end
```

> ¹Because `Phoenix.HTML` already provides these separate elements, In most Phoenix apps I've worked on; creating a component/function to abstract this "input group" has always been the general course of action to get this working.

**Comment by josevalim** [comment]
> As you said, we could add a format_value helper to Phoenix.HTML.Form. Another option that would work well is to move the formatting logic directly into the input_value function

The issue is that the valid value for the input depends on the type. So we either need to add a `type` option to input_value OR provide a different function. Given `input_value` is already in use today, and has its use cases, I would rather introduce a new function.

> (I'm also unsure what would be the use cases for a format_value function outside of this particular use-case).

To be clear, there are already two use cases for format_value. It can solve both 1 (format_input_value) and 2 (input_equals?). It solves the checkbox because it will always return true or false, which is easy to compare with.

> If we break the current generated .input helper apart, we can see that it's actually a group of label + input + error (perhaps we should even rename it to .input_group)¹.

It doesn't work. [I literally have tried that before](https://github.com/heartcombo/simple_form). At the end you end-up with so many configuration options to customize every degree of the generated markup for all JS/CSS use cases that it is not worth it IMO.

**Comment by josevalim** [comment]
To be clear, it doesn't work as a general framework abstraction. It works in your own apps because you only have to worry about _your_ use case. But Phoenix has to worry about _all_ use cases.

And even then, you have inputs that are slightly different and don't fit the mold, such as selects and checkboxes (which should typically be rendered inside labels). That's to say the current catch-all clause in `def input` _is_ your app `input_group`. :)

**Comment by thiagomajesk** [comment]
> It doesn't work. [I literally have tried that before](https://github.com/heartcombo/simple_form). At the end you end-up with so many configuration options to customize every degree of the generated markup for all JS/CSS use cases that it is not worth it IMO.

@josevalim I don't think you understood my proposal, the `input_group` helper is what you would get generated into the core components file. The individual functions are the ones that would be abstracted into `Phoenix.Component` :)

**Comment by josevalim** [comment]
@thiagomajesk right, I got that. My point is that I am not sure if that works. For example, for checkboxes, the input is inside the label, which means that the .input_group cannot have the label inside. It would rather be this:

```
def input_group(assigns) do
  ~H"""
    <div phx-feedback-for={@name}>
      <%= ...or a component %>
      <.errors let={error} form={f} field={field} translator={&translate_errors} class="phx-no-feedback:hidden mt-3 flex gap-3 text-sm leading-6 text-rose-600">
        <Heroicons.exclamation_circle mini class="mt-0.5 h-5 w-5 flex-none fill-rose-500" />
        <%= error %>
      </.errors>
    </div>
    """
end
```

And then, what is the point of making `.error` part of Phoenix.Component? Let's check its implementation:

```
  def error(assigns) do
    ~H"""
    <p class="phx-no-feedback:hidden mt-3 flex gap-3 text-sm leading-6 text-rose-600">
      <Heroicons.exclamation_circle mini class="mt-0.5 h-5 w-5 flex-none fill-rose-500" />
      <%= render_slot(@inner_block) %>
    </p>
    """
  end
```

It fully assumes you are using Tailwind (which Phoenix LiveView does not) and HeroIcons (same). If we all assumptions, we end up with:

```
  def error(assigns) do
    ~H"""
    <p class={@class}>
      <%= render_slot(@inner_block) %>
    </p>
    """
  end
```

And I doubt that's worth abstracting. The same for labels.

**Comment by josevalim** [comment]
With all that said, I think we could have a .input_group thingy in core_components for the feedback and error bit, but I don't think it help us encapsulate .input/.label/.error in Phoenix LV itself. Does that make sense? :)

**Comment by thiagomajesk** [comment]
I think that the common pattern to have `input`, `error` and `label` part of `Phoenix.Component` is to avoid having the code that extracts the values from the form inside the core components. The only drawback for the `input` component (like you mentioned) is in fact checkboxes. This is what I'm thinking right now:

 - `Phoenix.Component.label` will take the same args as `Phoenix.HTML.Form.label`
 - `Phoenix.Component.error` will extract the errors from the form and the translation logic
 - `Phoenix.Component.input` will abstract the calls to `Phoenix.HTML.Form.input_name/2`, `Phoenix.HTML.Form.input_id/2`, `Phoenix.HTML.Form.input_value/2`, `Phoenix.HTML.Form.options_for_select/2`
 
 Let's see if we can come up with a reasonable generic implementation that works for those cases (I'm gonna leave Tailwind out of the equation for simplicity's sake):
 
```elixir
def input_group(assigns) do
   ~H"""
    <div phx-feedback-for={@name}>
      <.label form={f} field={field} text={@label} />
      <.input type={@type} form={f} field={field} class={input_style(@type)} />
      <.errors let={error} form={f} field={field} translator={&translate_errors}>
        <p class={error_style()}><Heroicons.exclamation_circle mini /> <br /> <%= error %></p>
      </.errors>
    </div>
   """
end

def input_group(%{type: "checkbox"} = assigns) do
  ~H"""
    <label phx-feedback-for={@name} form={f} field={field}>
      <.input type={@type} form={f} field={field} class={input_style(@type)} />
      <.errors let={error} form={f} field={field} translator={&translate_errors}>
        <p class={error_style()}><Heroicons.exclamation_circle mini /> <br /> <%= error %></p>
      </.errors>
    </label>
  """
end
```

> And then, what is the point of making .error part of Phoenix.Component? Let's check its implementation:

Not just the component, but all it encompasses right now, calls to `translate_errors` and `translate_error` for instance. I honestly don't remember the last time I had to change these defaults in Phoenix, so it could be abstracted as a default implementation. I'm thinking about something along the lines of:

```elixir
def errors(assigns) do
   translate_fun = Map.get(assigns, :translator, &translate_errors)

   ~H"""
   <%= for error <- translate_fun.(@form.errors[@field]) do %>
      <%= render_slot(@inner_block, error) %>
    <% end %>
   """
end
```
 
Note that these aren't exact proposals, is just so we can get the conversation going and perhaps find a better solution to avoid having all this code in the core components (if we can). If this works, we could remove a lot of boilerplate code.

**Comment by josevalim** [comment]
> I think that the common pattern to have input, error and label part of Phoenix.Component is to avoid having the code that extracts the values from the form inside the core components.

I see. I have another solution for this problem, which is to implement `Access` for form, so we can do `form[field]` and that returns `Phoenix.HTML.FormInput` with all of name, id, errors, and value extracted out. This has the benefit that each we don't have to call the same `input_` functions over and over again on each component.

Even then, note that we cannot abstract label neither for checkboxes nor select (because they can be multiple which uses a slightly different naming schema).

> Not just the component, but all it encompasses right now, calls to translate_errors and translate_error for instance.

Your implementation is basically a enum map/for comprehension. I think it would be very hard to convince me this is worthy abstracting around.  :)

**Comment by thiagomajesk** [comment]
> I see. I have another solution for this problem, which is to implement `Access` for form, so we can do `form[field]` and that returns `Phoenix.HTML.FormInput` with all of name, id, errors, and value extracted out

This sounds interesting, could you share an example of what is in your mind?

> Even then, note that we cannot abstract label neither for checkboxes nor select (because they can be multiple which uses a slightly different naming schema)

The label component should have the `multiple` attr to account for the different naming logic, which would be very useful for other use cases where we want to generate grouped values (radios, checkboxes, selects, etc). I've lost track of how many times I wanted to do that and caught myself interpolating the names manually "form[field][]".

> Your implementation is basically a enum map/for comprehension. I think it would be very hard to convince me this is worthy abstracting around. :)

This is an oversimplification of the problem XD. In this case, the abstraction is not just around the comprehension, it also removes the unnecessary default translation code from the generators. Be honest... How frequently do you think those functions ever change in a standard Phoenix app? (Just to be sure I'm not talking from just my experience alone).

**Comment by josevalim** [comment]
The translation function assumes Gettext, which is not a dependency we can impose on Phoenix nor it is  a dependency of LiveView (and it can be readily opted out for new apps too). So you are asking us to bring a whole new dependency into LiveView to handle a concern that is app specific.

> Be honest... How frequently do you think those functions ever change in a standard Phoenix app? (Just to be sure I'm not talking from just my experience alone).

From my experience in trying to abstract forms in SimpleForm: it happens all the time. Even the smallest bit of UI will require customization.

**Comment by josevalim** [comment]
> The label component should have the multiple attr to account for the different naming logic, which would be very useful for other use cases where we want to generate grouped values (radios, checkboxes, selects, etc).

This is a good point but it also adds more complexity to the generated code because we now need to be passing an additional attribute to both label and input.

so in this case, the current code is better: this concern is handled once when we build the name and everything downstream works. This is another aspect to consider in whatever solution we come up with.

I will provider a proof of concept to what I have in mind later. :)

**Comment by thiagomajesk** [comment]
> From my experience in trying to abstract forms in SimpleForm: it happens all the time. Even the smallest bit of UI will require customization.

You're rightfully cautious, probably because of previous bad experiences with too much abstraction. But just to be clear: I was talking specifically about the translation code, which is just one entry point; not all things that might be customizable by the end-user, and from my experience, that translation code doesn't change very often. I have still to work on an app that uses something different than that default translation code (because those two functions basically just dispatch to Gettext).

> So you are asking us to bring a whole new dependency into LiveView to handle a concern that is app specific.

I looked very quickly at Phoenix's mix.exs and thought I saw gettext as a dependency, my mistake, it's for docs only. I don't think that's too app specific though, since Phoenix ships with this implementation by default, and as I said, I don't think I ever have changed it to do something else. But I agree that bringing a new dependency might not be desirable for some reasons. I also completely agree with your concern that Phoenix should keep 100% UI agnostic.
